# Setup MySQL Database

Truth Social dataset were presented by [Truth Social Dataset](https://arxiv.org/abs/2303.11240) paper and can be downloaded [HERE](https://zenodo.org/record/7531625)

## Imports

In [31]:
#!pip install pymysql

  Obtaining dependency information for pymysql from https://files.pythonhosted.org/packages/e5/30/20467e39523d0cfc2b6227902d3687a16364307260c75e6a1cb4422b0c62/PyMySQL-1.1.0-py3-none-any.whl.metadata
   ---------------------------------------- 44.8/44.8 kB 2.2 MB/s eta 0:00:00


In [2]:
import mysql.connector
import pandas as pd
from sqlalchemy import create_engine
import pymysql

## DB Connection

In [3]:
psw = '' # Password

conn = mysql.connector.connect(
    host = 'localhost',
    user = 'root', 
    password = psw,
    database = 'truth_social')

cursor = conn.cursor()

In [4]:
db_conn = f"mysql+pymysql://root:{psw}@localhost/truth_social"
engine = create_engine(db_conn)

## Data cleaning and CRUD 

### Users.tsv

* Null values are represented as -1 
* 432503 users have NULL values on ```timestamp```, ```follower_count```, ```following_count```, ```finished_follower_scrape```, ```finished_following_scrape``` and ```finished_truth_scrape``` columns.
* 18 users have NULL values on ```username``` field

In [3]:
df = pd.read_csv('truth_social/users.tsv', sep='\t', low_memory=False)
df.head()

id            timestamp                time_scraped      username  \
0  1238  2022-02-01 00:00:00   2022-09-19 14:43:26.11971     jeffgolfs   
1    50  2022-02-01 00:00:00  2022-09-15 02:01:38.652706          MizW   
2  6951  2022-02-01 00:00:00  2022-10-08 17:23:33.822145   SusanCreech   
3  9861  2022-02-01 00:00:00   2022-10-16 20:43:09.71727  BMF6FeeTDeeP   
4    34  2022-02-01 00:00:00  2022-09-14 23:15:35.220658     DC_Draino   

   follower_count  following_count                            profile_url  \
0             886              204     https://truthsocial.com/@jeffgolfs   
1            2365             2929          https://truthsocial.com/@MizW   
2             870             1232   https://truthsocial.com/@SusanCreech   
3            1086             1654  https://truthsocial.com/@BMF6FeeTDeeP   
4          286146               81     https://truthsocial.com/@DC_Draino   

  finished_follower_scrape finished_following_scrape finished_truth_scrape  
0                        f                         f                     f  
1                        f                         f                     f  
2                        f                         f                     f  
3                        f                         f                     f  
4                        t                         t                     t

In [4]:
null_count = df.applymap(lambda x: x == -1 or x == '-1').sum()
null_count

id                                0
timestamp                    432503
time_scraped                      0
username                         18
follower_count               432503
following_count              432503
profile_url                       0
finished_follower_scrape     432503
finished_following_scrape    432503
finished_truth_scrape        432503
dtype: int64

In [5]:
df[df['username'] == '-1'].head()

id timestamp                time_scraped username  follower_count  \
119490  119560        -1  2022-12-23 09:08:18.406512       -1              -1   
127811  127881        -1  2022-12-23 09:40:12.856304       -1              -1   
178390  178460        -1  2022-12-23 13:47:00.219298       -1              -1   
210208  210278        -1  2022-12-23 17:03:27.756373       -1              -1   
275812  275882        -1  2022-12-24 01:27:00.564675       -1              -1   

        following_count                   profile_url  \
119490               -1  https://truthsocial.com/@nan   
127811               -1  https://truthsocial.com/@nan   
178390               -1  https://truthsocial.com/@nan   
210208               -1  https://truthsocial.com/@nan   
275812               -1  https://truthsocial.com/@nan   

       finished_follower_scrape finished_following_scrape  \
119490                       -1                        -1   
127811                       -1                        -1   
178390                       -1                        -1   
210208                       -1                        -1   
275812                       -1                        -1   

       finished_truth_scrape  
119490                    -1  
127811                    -1  
178390                    -1  
210208                    -1  
275812                    -1

In [18]:
# Replace NULL values (-1) with None. 
# Set Boolean columns to True or False

df = df.replace(-1, None)
df = df.replace('-1', None)
boolean_columns = df.columns[-3:]
df[boolean_columns] = df[boolean_columns].replace({'t': True, 'f': False})

Creating and populating table

In [31]:
command = """CREATE TABLE IF NOT EXISTS users(
            id int NOT NULL PRIMARY KEY,
            timestamp datetime,
            time_scraped datetime,
            username varchar(255),
            follower_count int,
            following_count int,
            profile_url varchar(512),
            finished_follower_scrape bool,
            finished_following_scrape bool,
            finished_truth_scrape bool
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [36]:
df.to_sql('users', engine, if_exists='append', index=False)

454458

### Follows.tsv

* ```followed``` field has 1 id that is not on ```users``` table (id = 97). The corresponding row was deleted

In [37]:
df = pd.read_csv('truth_social/follows.tsv', sep='\t', low_memory=False)
df.head()

id                time_scraped  follower  followed
0  2701620     2022-11-03 18:55:20.902     12475     22025
1  2701622  2022-11-03 18:55:22.675183     12475     22026
2  2701624  2022-11-03 18:55:23.910835     12475     22027
3  2701626  2022-11-03 18:55:25.455047     12475     22028
4  2701628  2022-11-03 18:55:27.146986     12475     17280

In [54]:
users = pd.read_csv('truth_social/users.tsv', sep='\t', low_memory=False)
users.head()

id            timestamp                time_scraped      username  \
0  1238  2022-02-01 00:00:00   2022-09-19 14:43:26.11971     jeffgolfs   
1    50  2022-02-01 00:00:00  2022-09-15 02:01:38.652706          MizW   
2  6951  2022-02-01 00:00:00  2022-10-08 17:23:33.822145   SusanCreech   
3  9861  2022-02-01 00:00:00   2022-10-16 20:43:09.71727  BMF6FeeTDeeP   
4    34  2022-02-01 00:00:00  2022-09-14 23:15:35.220658     DC_Draino   

   follower_count  following_count                            profile_url  \
0             886              204     https://truthsocial.com/@jeffgolfs   
1            2365             2929          https://truthsocial.com/@MizW   
2             870             1232   https://truthsocial.com/@SusanCreech   
3            1086             1654  https://truthsocial.com/@BMF6FeeTDeeP   
4          286146               81     https://truthsocial.com/@DC_Draino   

  finished_follower_scrape finished_following_scrape finished_truth_scrape  
0                        f                         f                     f  
1                        f                         f                     f  
2                        f                         f                     f  
3                        f                         f                     f  
4                        t                         t                     t

In [56]:
follower_ids = df['follower'].unique()
followed_ids = df['followed'].unique()

user_ids = users['id'].unique()

is_follower_valid = all(x in user_ids for x in follower_ids)
is_followed_valid = all(x in user_ids for x in followed_ids)

In [57]:
is_follower_valid

True

In [58]:
is_followed_valid

False

In [59]:
not_in_users = [x for x in followed_ids if x not in user_ids]
len(not_in_users)

1

In [60]:
not_in_users

[97]

In [65]:
# Removing 1 row with invalid user id

df = df.drop(df[df.followed == 97].index)

Creating and populating table

In [49]:
command = """CREATE TABLE IF NOT EXISTS follows(
            id int NOT NULL PRIMARY KEY,
            time_scraped datetime,
            follower int,
            followed int,
            FOREIGN KEY (follower) REFERENCES users(id),
            FOREIGN KEY (followed) REFERENCES users(id)
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [66]:
df.to_sql('follows', engine, if_exists='append', index=False)

4002114

### Truths.tsv

* The file has 854,432 rows, but the author's documentation says it has 845,060.
* There are some problems with the tsv format. Some texts have symbols interpreted as tab, which gives more fields. From 854,432 instances, 739,779 are in the correct format
* Encountered errors: 
    * field larger than field limit (131072)
    * '	' expected after '"'
    * unexpected end of data
    * Expected 13 fields in line 16669, saw 15
    * Expected 13 fields in line 16669, saw 14
* The rows with 2 exceeding fields have the author's name after ```text```field and an empty field after ```post_url``` field
* The rows with 1 exceeding field have the author's name after ```text```field
* Those errors were fixed
* The row 519,948 has a weird format, with multiple fields. It was ignored.
* 237 truths without author's id. They were ignored
* NaN values on ```truth_retruthed``` field were replaced by -2. -1 values remained unchanged.
* 4 author's id were not present on ```users``` table. They were deleted
* There are 12,236 duplicated lines. Only 1 copy was kept.
* The final dataframe had 837,979 instances

In [65]:
def process_bad_lines(bad_line: list[str]) -> list[str]:
    if len(bad_line) == 15:
        bad_line.pop(10)
        bad_line.pop(13)
        return bad_line
    elif len(bad_line) == 14:
        bad_line.pop(10)
        return bad_line
    print(bad_line)
    return None

In [66]:
df = pd.read_csv('truth_social/truths.tsv', sep='\t', on_bad_lines=process_bad_lines, engine="python", quoting=3)
df.head()

['430702', '2022-03-15 05:47:00', '2022-10-14 22:34:56.932013', 'f', 'f', '9375', '0', '0', '0', '"“Ming/Putin” (Klytus/Biden) I\'\'m bored. What plaything can you offer me today?)\u2028430709', '2022-03-07 07:16:00', '2022-10-14 22:35:25.809822', 't', 'f', '9375', '0', '0', '0', 'The Russian people are struggling but this is funny.', '107926893084469163', 'https://truthsocial.com/@jmzucco/posts/107926893084469163', '983689']


id            timestamp                time_scraped is_retruth  \
0  703265  2022-05-14 00:00:00  2022-11-01 15:48:16.829245          t   
1  807614  2022-03-27 00:00:00  2022-11-23 10:54:36.865046          t   
2  807615  2022-03-29 00:00:00  2022-11-23 10:55:11.613249          t   
3  807618  2022-03-29 00:00:00  2022-11-23 10:56:29.134825          t   
4  807619  2022-03-29 00:00:00  2022-11-23 10:57:20.028953          f   

  is_reply   author  like_count  retruth_count  reply_count  \
0        f   2247.0           0              0            0   
1        f  20054.0           0              0            0   
2        f  20054.0           0              0            0   
3        f  20054.0           0              0            0   
4        t  20054.0           0              0            0   

                                                text         external_id  \
0  Q+ BE READY ANONS - PUBLIC AWAKENING COMING - ...  108351978111014587   
1                          Enough is enough! RETRUTH  108040697775316106   
2  https://justthenews.com/politics-policy/all-th...  108040478595657472   
3                    https://t.me/realx22report/6729  108040420546204424   
4                                     @CeceBloomwood  108040224733670356   

                                                 url  truth_retruthed  
0  https://truthsocial.com/@jweimer39/posts/10835...        1043583.0  
1  https://truthsocial.com/@CeceBloomwood/posts/1...         988567.0  
2  https://truthsocial.com/@CeceBloomwood/posts/1...        1002825.0  
3  https://truthsocial.com/@CeceBloomwood/posts/1...         989757.0  
4  https://truthsocial.com/@CeceBloomwood/posts/1...             -1.0

In [67]:
df.shape

(854431, 13)

In [68]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 854431 entries, 0 to 854430
Data columns (total 13 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   id               854431 non-null  int64  
 1   timestamp        854431 non-null  object 
 2   time_scraped     854431 non-null  object 
 3   is_retruth       854431 non-null  object 
 4   is_reply         854431 non-null  object 
 5   author           854194 non-null  float64
 6   like_count       854431 non-null  int64  
 7   retruth_count    854431 non-null  int64  
 8   reply_count      854431 non-null  int64  
 9   text             854113 non-null  object 
 10  external_id      854431 non-null  int64  
 11  url              854431 non-null  object 
 12  truth_retruthed  825804 non-null  float64
dtypes: float64(2), int64(5), object(6)
memory usage: 84.7+ MB


In [69]:
null_count = df.applymap(lambda x: x == -1 or x == '-1').sum()
null_count

id                      0
timestamp           81194
time_scraped            0
is_retruth              0
is_reply                0
author                  0
like_count              0
retruth_count           0
reply_count             0
text                    0
external_id             0
url                     0
truth_retruthed    586506
dtype: int64

In [80]:
df.loc[df['truth_retruthed'].isna()]

id            timestamp                time_scraped is_retruth  \
1165    807653  2022-03-12 00:00:00  2022-11-23 11:11:34.708273          t   
1166    807654  2022-02-14 00:00:00  2022-11-23 11:12:07.513646          t   
1167    807655  2022-03-07 00:00:00  2022-11-23 11:12:36.069335          t   
1168    807656  2022-03-05 00:00:00  2022-11-23 11:13:06.035362          t   
1169    807657  2022-03-04 00:00:00  2022-11-23 11:13:31.823003          t   
...        ...                  ...                         ...        ...   
854426  665713  2022-08-26 16:23:00  2022-10-26 07:34:31.909649          f   
854427  665613  2022-09-23 00:00:00  2022-10-26 07:23:49.617289          t   
854428  665615  2022-09-22 00:00:00  2022-10-26 07:24:10.213432          f   
854429  665672  2022-09-04 00:00:00  2022-10-26 07:30:23.533308          t   
854430  665679  2022-08-26 20:07:00  2022-10-26 07:31:02.113298          f   

       is_reply  author  like_count  retruth_count  reply_count  \
1165          f   20054           0              0            0   
1166          f   20054      269000          51000        37500   
1167          f   20054         615            378           21   
1168          f   20054        1300            526           81   
1169          f   20054        1260            692           45   
...         ...     ...         ...            ...          ...   
854426        t   12967           1              0            0   
854427        f   13173         213            119           42   
854428        f   13173           2              0            0   
854429        f   13173           6              4            1   
854430        t   12967           0              0            0   

                                                     text         external_id  \
1165    "TRUMP: “Before we can defeat the Democrats, S...  107949547160373982   
1166    Get Ready!  Your favorite President will see y...  107945338731779453   
1167                                                  NaN  107925568586381393   
1168    We have people running the federal government ...  107914242452173533   
1169                                               Truth!  107912178895243526   
...                                                   ...                 ...   
854426  I found a bad troll that you might be interest...  108891040156240143   
854427                                                  <  109046056601070160   
854428  https://www.coindesk.com/policy/2022/09/21/sec...  109042598301559137   
854429  woooo hooooooo FREEEEEDOM <emoji: cupid><emoji...  108942968601431628   
854430                            <emoji: +1><emoji: 100>  108891920615147716   

                                                      url  truth_retruthed  
1165    https://truthsocial.com/@CeceBloomwood/posts/1...              NaN  
1166    https://truthsocial.com/@CeceBloomwood/posts/1...              NaN  
1167    https://truthsocial.com/@CeceBloomwood/posts/1...              NaN  
1168    https://truthsocial.com/@CeceBloomwood/posts/1...              NaN  
1169    https://truthsocial.com/@CeceBloomwood/posts/1...              NaN  
...                                                   ...              ...  
854426  https://truthsocial.com/@TUBA1956/posts/108891...              NaN  
854427  https://truthsocial.com/@johnnygromax/posts/10...              NaN  
854428  https://truthsocial.com/@johnnygromax/posts/10...              NaN  
854429  https://truthsocial.com/@johnnygromax/posts/10...              NaN  
854430  https://truthsocial.com/@TUBA1956/posts/108891...              NaN  

[28390 rows x 13 columns]

In [78]:
df.loc[df['author'].isna()]

id            timestamp                time_scraped is_retruth  \
134639  895511  2022-08-19 00:00:00  2022-12-09 07:10:54.473982          f   
134654  895527  2022-08-14 00:00:00  2022-12-09 07:11:59.357284          f   
134663  895535  2022-08-11 00:00:00  2022-12-09 07:12:25.754131          f   
134664  895536  2022-09-25 00:00:00  2022-12-09 07:12:27.997297          f   
134679  895573  2022-07-13 00:00:00  2022-12-09 07:15:19.227475          f   
...        ...                  ...                         ...        ...   
851753  942094  2022-08-22 00:00:00  2022-12-12 08:16:13.174997          f   
851755  942098  2022-06-22 00:00:00  2022-12-12 08:16:27.227282          f   
851760  942093  2022-06-19 00:00:00  2022-12-12 08:16:10.239296          f   
851762  942097  2022-08-22 00:00:00  2022-12-12 08:16:25.206955          f   
851768  942106  2022-08-20 00:00:00  2022-12-12 08:17:10.644137          f   

       is_reply  author  like_count  retruth_count  reply_count  \
134639        f     NaN          55             31            6   
134654        f     NaN        6030           3020          332   
134663        f     NaN        1870            786          282   
134664        f     NaN       10000           2530          298   
134679        f     NaN        4240           1380          244   
...         ...     ...         ...            ...          ...   
851753        f     NaN        6410           1430          279   
851755        f     NaN          32             30            1   
851760        f     NaN           3              0            0   
851762        f     NaN        1600            305           43   
851768        f     NaN        2350            394          106   

                                                     text         external_id  \
134639                                                  <  108864982160272342   
134654  Obama and Hillary were running guns from Libya...  108822381542261053   
134663                                                  <  108814846897812544   
134664                                                  <  109073924301696705   
134679  Did John Bolton realize he just publicly admit...  108639442992116282   
...                                                   ...                 ...   
851753  We should seize Fauci’s passport since we’re d...  108867581547173281   
851755                                                  <  108659435882915641   
851760                                                  <  108669737640159756   
851762  These AI art bots are getting so good they’re ...  108867048287366701   
851768  There’s been a bunch of ‘legacy celebs’ like A...  108856308667149579   

                                                      url  truth_retruthed  
134639  https://truthsocial.com/@truthtent/posts/10886...              NaN  
134654  https://truthsocial.com/@JackPosobiec/posts/10...              NaN  
134663  https://truthsocial.com/@JohnPaulMacIsaac/post...              NaN  
134664  https://truthsocial.com/@Lara45/posts/10907392...              NaN  
134679  https://truthsocial.com/@JackPosobiec/posts/10...              NaN  
...                                                   ...              ...  
851753  https://truthsocial.com/@JackPosobiec/posts/10...              NaN  
851755  https://truthsocial.com/@Mercyofangel/posts/10...              NaN  
851760  https://truthsocial.com/@Mathed/posts/10866973...              NaN  
851762  https://truthsocial.com/@JackPosobiec/posts/10...              NaN  
851768  https://truthsocial.com/@JackPosobiec/posts/10...              NaN  

[237 rows x 13 columns]

In [82]:
# Convert float columns to int
# Set Boolean columns
# Replace Null values (-1) with None

df.dropna(subset=['author'], inplace=True)
df['truth_retruthed'].fillna(-2, inplace=True)

df['author'] = df['author'].astype(int)
df['truth_retruthed'] = df['truth_retruthed'].astype(int)

boolean_columns = ['is_retruth', 'is_reply']
df[boolean_columns] = df[boolean_columns].replace({'t': True, 'f': False})

df['timestamp'] = df['timestamp'].replace('-1', None)

In [83]:
df

id            timestamp                time_scraped  is_retruth  \
0       703265  2022-05-14 00:00:00  2022-11-01 15:48:16.829245        True   
1       807614  2022-03-27 00:00:00  2022-11-23 10:54:36.865046        True   
2       807615  2022-03-29 00:00:00  2022-11-23 10:55:11.613249        True   
3       807618  2022-03-29 00:00:00  2022-11-23 10:56:29.134825        True   
4       807619  2022-03-29 00:00:00  2022-11-23 10:57:20.028953       False   
...        ...                  ...                         ...         ...   
854426  665713  2022-08-26 16:23:00  2022-10-26 07:34:31.909649       False   
854427  665613  2022-09-23 00:00:00  2022-10-26 07:23:49.617289        True   
854428  665615  2022-09-22 00:00:00  2022-10-26 07:24:10.213432       False   
854429  665672  2022-09-04 00:00:00  2022-10-26 07:30:23.533308        True   
854430  665679  2022-08-26 20:07:00  2022-10-26 07:31:02.113298       False   

        is_reply  author  like_count  retruth_count  reply_count  \
0          False    2247           0              0            0   
1          False   20054           0              0            0   
2          False   20054           0              0            0   
3          False   20054           0              0            0   
4           True   20054           0              0            0   
...          ...     ...         ...            ...          ...   
854426      True   12967           1              0            0   
854427     False   13173         213            119           42   
854428     False   13173           2              0            0   
854429     False   13173           6              4            1   
854430      True   12967           0              0            0   

                                                     text         external_id  \
0       Q+ BE READY ANONS - PUBLIC AWAKENING COMING - ...  108351978111014587   
1                               Enough is enough! RETRUTH  108040697775316106   
2       https://justthenews.com/politics-policy/all-th...  108040478595657472   
3                         https://t.me/realx22report/6729  108040420546204424   
4                                          @CeceBloomwood  108040224733670356   
...                                                   ...                 ...   
854426  I found a bad troll that you might be interest...  108891040156240143   
854427                                                  <  109046056601070160   
854428  https://www.coindesk.com/policy/2022/09/21/sec...  109042598301559137   
854429  woooo hooooooo FREEEEEDOM <emoji: cupid><emoji...  108942968601431628   
854430                            <emoji: +1><emoji: 100>  108891920615147716   

                                                      url  truth_retruthed  
0       https://truthsocial.com/@jweimer39/posts/10835...          1043583  
1       https://truthsocial.com/@CeceBloomwood/posts/1...           988567  
2       https://truthsocial.com/@CeceBloomwood/posts/1...          1002825  
3       https://truthsocial.com/@CeceBloomwood/posts/1...           989757  
4       https://truthsocial.com/@CeceBloomwood/posts/1...               -1  
...                                                   ...              ...  
854426  https://truthsocial.com/@TUBA1956/posts/108891...               -2  
854427  https://truthsocial.com/@johnnygromax/posts/10...               -2  
854428  https://truthsocial.com/@johnnygromax/posts/10...               -2  
854429  https://truthsocial.com/@johnnygromax/posts/10...               -2  
854430  https://truthsocial.com/@TUBA1956/posts/108891...               -2  

[854194 rows x 13 columns]

In [108]:
users = pd.read_csv('truth_social/users.tsv', sep='\t', low_memory=False)

In [106]:
author_ids = df['author'].unique()
user_ids = users['id'].unique()

not_in_users = [x for x in author_ids if x not in user_ids]
len(not_in_users)

4

In [107]:
not_in_users

[6364, 97, 7838, 1173]

In [109]:
df = df[~df['author'].isin(not_in_users)]
df.shape

(850215, 13)

In [122]:
df = df.drop_duplicates(subset=['id'])

In [123]:
df.shape

(837979, 13)

Creating and populating table

In [90]:
command = """CREATE TABLE IF NOT EXISTS truths(
            id int NOT NULL PRIMARY KEY,
            timestamp datetime,
            time_scraped datetime,
            is_retruth bool, 
            is_reply bool,
            author int,
            like_count int,
            retruth_count int,
            reply_count int,
            text varchar(14000),
            external_id int,
            url varchar(255),
            truth_retruthed int,
            FOREIGN KEY (author) REFERENCES users(id)
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [101]:
command = "ALTER TABLE truths DROP COLUMN external_id;"
cursor.execute(command)
conn.commit()  # edit db 

In [102]:
command = "ALTER TABLE truths ADD COLUMN external_id varchar(50)"
cursor.execute(command)
conn.commit()  # edit db

In [126]:
df.to_sql('truths', engine, if_exists='append', index=False)

837979

### Replies.tsv

* 4 replying_users are not on ```users``` table. They were deleted (ids = [6364, 97, 7838, 1173])

In [5]:
df = pd.read_csv('truth_social/replies.tsv', sep='\t', low_memory=False)
df.head()

id                time_scraped  replying_user  replied_user
0  470149  2022-12-11 02:50:32.813591            536           708
1       4  2022-09-15 23:06:01.361947            177         68826
2       5  2022-09-15 23:06:09.480978            177          9281
3       6  2022-09-15 23:06:10.593096            177          6499
4   13151  2022-09-26 00:39:29.734157           1549         15879

In [11]:
df.shape

(506276, 4)

In [8]:
users = pd.read_csv('truth_social/users.tsv', sep='\t', low_memory=False)

In [9]:
replying_ids = df['replying_user'].unique()
replied_ids = df['replied_user'].unique()
user_ids = users['id'].unique()

invalid_replying = [x for x in replying_ids if x not in user_ids]
invalid_replied = [x for x in replied_ids if x not in user_ids]

print(f'Invalid replying users: {len(invalid_replying)}')
print(f'Invalid replied users: {len(invalid_replied)}')

Invalid replying users: 4
Invalid replied users: 0


In [10]:
invalid_replying

[6364, 97, 7838, 1173]

In [13]:
df = df[~df['replying_user'].isin(invalid_replying)]
df.shape

(505945, 4)

Creating and populating table

In [14]:
command = """CREATE TABLE IF NOT EXISTS replies(
            id int NOT NULL PRIMARY KEY,
            time_scraped datetime,
            replying_user int,
            replied_user int,
            FOREIGN KEY (replying_user) references users(id),
            FOREIGN KEY (replied_user) references users(id)
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [15]:
df.to_sql('replies', engine, if_exists='append', index=False)

505945

### Quotes.tsv

* 914 invalid quoting_truths ids. They were ignored.

In [16]:
df = pd.read_csv('truth_social/quotes.tsv', sep='\t', low_memory=False)
df.head()

id            timestamp                time_scraped  quoted_user  \
0    197  2022-07-22 09:43:00  2022-09-17 04:57:35.624255          125   
1    181  2022-08-20 18:00:00  2022-09-17 03:49:16.168681          300   
2   3492  2022-07-14 21:13:00  2022-10-13 02:18:43.045302         2106   
3     10  2022-09-12 22:40:00  2022-09-15 22:37:05.004004          933   
4  10489  2022-11-11 00:00:00  2022-11-13 23:44:24.533708       134003   

   quoting_user  quoting_truth  quoted_truth_external_id  
0           131          16333        108688461672926241  
1           300          14949        108857444588213255  
2          8212         367606        108648254976307349  
3            27           1593        108790130246424349  
4          6657         763581        109324987069121051

In [20]:
df.shape

(10508, 7)

In [ ]:
users = pd.read_csv('truth_social/users.tsv', sep='\t', low_memory=False)

In [19]:
quoting_users = df['quoting_user'].unique()
quoted_users = df['quoted_user'].unique()
user_ids = users['id'].unique()

invalid_quoting = [x for x in quoting_users if x not in user_ids]
invalid_quoted = [x for x in quoted_users if x not in user_ids]

print(f'Invalid quoting users: {len(invalid_quoting)}')
print(f'Invalid quoted users: {len(invalid_quoted)}')

Invalid quoting users: 1
Invalid quoted users: 0


In [21]:
df = df[~df['quoting_user'].isin(invalid_quoting)]
df.shape

(10499, 7)

In [36]:
command = "SELECT id from truths"
cursor.execute(command)
truths = cursor.fetchall()
truths_ids = [x[0] for x in truths]

In [38]:
quoting_truths = df['quoting_truth'].unique()
invalid_quoting_truths = [x for x in quoting_truths if x not in truths_ids]
len(invalid_quoting_truths)

914

In [40]:
df = df[~df['quoting_truth'].isin(invalid_quoting_truths)]

In [41]:
df.shape

(9585, 7)

Creating and populating table

In [23]:
command = """CREATE TABLE IF NOT EXISTS quotes(
            id int NOT NULL PRIMARY KEY,
            timestamp datetime,
            time_scraped datetime,
            quoted_user int,
            quoting_user int,
            quoting_truth int,
            quoted_truth_external_id varchar(50),
            FOREIGN KEY (quoted_user) references users(id),
            FOREIGN KEY (quoting_user) references users(id),
            FOREIGN KEY (quoting_truth) references truths(id)
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [42]:
df.to_sql('quotes', engine, if_exists='append', index=False)

9585

### Truth_user_tag_edges.tsv

* 47 invalid user_ids and 10,912 invalid truth_ids. Because of big numbers, they were kept without being foreign keys to ```truths``` and ```users``` tables

In [51]:
df = pd.read_csv('truth_social/truth_user_tag_edges.tsv', sep='\t', low_memory=False)
df.head()

id                time_scraped  truth_id  user_id
0   1  2022-09-14 20:55:51.203708        21       32
1   2  2022-09-14 23:15:38.465849        53       34
2   3  2022-09-15 01:51:27.835997        57       27
3   4  2022-09-15 01:51:33.178041        63       27
4   5  2022-09-15 01:51:39.765255        67       36

In [45]:
invalid_users = [x for x in df['user_id'].unique() if x not in user_ids]
len(invalid_users)

47

In [46]:
invalid_truths = [x for x in df['truth_id'].unique() if x not in truths_ids]
len(invalid_truths)

10912

In [49]:
#df = df[~df['truth_id'].isin(invalid_truths)]
#df = df[~df['user_id'].isin(invalid_users)]

In [52]:
df.shape

(145234, 4)

Creating and populating table

In [53]:
command = """CREATE TABLE IF NOT EXISTS truth_user_tag_edges(
            id int NOT NULL PRIMARY KEY,
            time_scraped datetime,
            truth_id int,
            user_id int
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [54]:
df.to_sql('truth_user_tag_edges', engine, if_exists='append', index=False)

145234

### media.tsv

In [55]:
df = pd.read_csv('truth_social/media.tsv', sep='\t', low_memory=False)
df.head()

id                                          media_url
0   2  https://static-assets-1.truthsocial.com/tmtg:p...
1   5  https://static-assets-1.truthsocial.com/tmtg:p...
2   6  https://static-assets-1.truthsocial.com/tmtg:p...
3   8  https://static-assets-1.truthsocial.com/tmtg:p...
4   9  https://static-assets-1.truthsocial.com/tmtg:p...

In [57]:
df['media_url'].str.len().max()

835

In [58]:
command = """CREATE TABLE IF NOT EXISTS media(
            id int NOT NULL PRIMARY KEY,
            media_url varchar(1000)
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [59]:
df.to_sql('media', engine, if_exists='append', index=False)

184884

### Truth_media_edges.tsv

* ```truth_id``` is not a foreign key

In [60]:
df = pd.read_csv('truth_social/truth_media_edges.tsv', sep='\t', low_memory=False)
df.head()

id  truth_id  media_id
0   2        37         2
1   5        40         5
2   6        42         6
3   8        44         8
4   9        45         9

In [61]:
command = """CREATE TABLE IF NOT EXISTS truth_media_edges(
            id int NOT NULL PRIMARY KEY,
            truth_id int,
            media_id int,
            FOREIGN KEY (media_id) REFERENCES media(id)
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [62]:
df.to_sql('truth_media_edges', engine, if_exists='append', index=False)

257500

### Hashtags.tsv

In [63]:
df = pd.read_csv('truth_social/hashtags.tsv', sep='\t', low_memory=False)
df.head()

id                  first_seen         hashtag
0   1  2022-09-14 17:29:14.578577       MarkGeist
1   2  2022-09-14 17:34:04.180279         LIVGolf
2   3  2022-09-14 23:15:11.267669  TrumpTurnberry
3   4  2022-09-14 23:15:12.117912   TrumpAberdeen
4   5  2022-09-14 23:15:39.463634          Alaska

In [65]:
df['hashtag'].str.len().max()

351

In [66]:
command = """CREATE TABLE IF NOT EXISTS hashtags(
            id int NOT NULL PRIMARY KEY,
            first_seen datetime,
            hashtag varchar(500)
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [67]:
df.to_sql('hashtags', engine, if_exists='append', index=False)

21599

### Truth_hashtag_edges.tsv

* ```truth_id``` is not a foreign key
* 8000+ invalid hashtag ids. They were ignored

In [68]:
df = pd.read_csv('truth_social/truth_hashtag_edges.tsv', sep='\t', low_memory=False)
df.head()

id  truth_id  hashtag_id
0   1        18           1
1   4        21           2
2   5        37           3
3   6        37           4
4   7        54           5

In [73]:
cursor.execute("SELECT id FROM hashtags")
hashtag_ids = cursor.fetchall()
hashtag_ids = [x[0] for x in hashtag_ids]
len(hashtag_ids)

21599

In [76]:
df = df[df['hashtag_id'].isin(hashtag_ids)]
df.shape

(205214, 3)

In [69]:
command = """CREATE TABLE IF NOT EXISTS truth_hashtag_edges(
            id int NOT NULL PRIMARY KEY,
            truth_id int,
            hashtag_id int, 
            FOREIGN KEY (hashtag_id) REFERENCES hashtags(id)
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [77]:
df.to_sql('truth_hashtag_edges', engine, if_exists='append', index=False)

205214

### External_urls.tsv

In [78]:
df = pd.read_csv('truth_social/external_urls.tsv', sep='\t', low_memory=False)
df.head()

id                                                url
0  1780  https://www.supremecourt.gov/opinions/21pdf/19...
1     3  https://nypost.com/2022/08/29/fbi-agent-resign...
2     4  https://brave-books-series.myshopify.com/produ...
3     5  https://www.foxnews.com/politics/polls-now-clo...
4     6  https://nypost.com/2022/04/17/the-verdict-is-i...

In [79]:
df['url'].str.len().max()

606

In [80]:
command = """CREATE TABLE IF NOT EXISTS external_urls(
            id int NOT NULL PRIMARY KEY,
            url varchar(700)
            )"""

cursor.execute(command)
conn.commit()  # edit db 

In [81]:
df.to_sql('external_urls', engine, if_exists='append', index=False)

173947

### Truth_external_url_edges.tsv

In [82]:
df = pd.read_csv('truth_social/truth_external_url_edges.tsv', sep='\t', low_memory=False)
df.head()

id  truth_id  url_id
0   4        41       3
1   5        83       4
2   6        97       5
3   7       108       6
4   8       115       7

In [83]:
command = """CREATE TABLE IF NOT EXISTS external_urls(
            id int NOT NULL PRIMARY KEY,
            truth_id int,
            url_id int,
            FOREIGN KEY (url_id) REFERENCES external_urls(id)
            )"""

cursor.execute(command)
conn.commit()  # edit db 

### Close connection

In [67]:
cursor.close()
conn.close()